### ISSUE TITLE
Any recommend way to Liveness built-in detection?

### ISSUE THREAD

**Issue by Ahmedsaber9** [exadel-inc/CompreFace#802]
Hello
I want to know if the detection (The user should perform specific eye. mouth, or head movements)
is that supported or could be implemented if any has any idea to get that done I'm open to knowing.

Thanks

**Comment by Ahmedsaber9** [comment]
@pospielov

**Comment by pospielov** [comment]
It can be implemented using the `landmarks` plugin. 
using it, you can check what the current head pose is.
We don't have an article about it, but here is an example of how amazon recommends implementing it using their facial recognition service: 
https://aws.amazon.com/blogs/industries/liveness-detection-to-improve-fraud-prevention-in-financial-institutions-with-amazon-rekognition/
There is an even more powerful plugin `landmarks2d106` that returns 106 points. Using it you can check if the mouth or eyes are opened.

**Comment by Ahmedsaber9** [comment]
Hi @pospielov 

I take a look at https://github.com/deepinsight/insightface/tree/ce3600a74209808017deaf73c036759b96a44ccb/alignment/coordinate_reg#visualization 
that mentioned at https://github.com/exadel-inc/CompreFace/blob/master/docs/Face-services-and-plugins.md#face-plugins
could you provide guidance to make that work with CompreFace API ? as mentioned it's require some manual config.

Tested Req

![image](https://user-images.githubusercontent.com/23035531/176981576-08fb286f-317f-4c58-a429-99247c3b5615.png)

Thanks.

**Comment by Ahmedsaber9** [comment]
much thanks, 
What about the Livens challenges ?
I took a look at what you shared with me "https://aws.amazon.com/blogs/industries/liveness-detection-to-improve-fraud-prevention-in-financial-institutions-with-amazon-rekognition/"

but the whole code is closeted to amazon infrastructure, do you recommend any project or something like that to integrate the face livens test?

**Comment by pospielov** [comment]
Unfortunately, I don't know about such projects. I am afraid you need to implement it by yourself, I believe it's possible to make it work locally, without amazon infrastructure

**Comment by Ahmedsaber9** [comment]
I'm working on using your build, but I need some help, 
first, I run docker pull pospielov/compreface-core:1.0.0-mobilenet-2d106

"I have CompreFace face already installed and running with the default docker image 5 landmarks at the same server using my public IP and everything is working fine" .

I tried with :

`docker-compose up -d  pospielov/compreface-core:1.0.0-mobilenet-2d106
`
but 

```
docker-compose up -d  pospielov/compreface-core:1.0.0-mobilenet-2d106
Creating network "word_default" with the default driver
Creating volume "word_postgres-data" with the default driver
ERROR: No such service: pospielov/compreface-core:1.0.0-mobilenet-2d106

```
also i tried with ` docker run -d -p 8080:80 pospielov/compreface-core:1.0.0-mobilenet-2d106`

but when I try to use my server IP it has an `ERR_CONNECTION_REFUSED`

sorry I'm not so experienced with the docker network 

Thanks

**Comment by Ahmedsaber9** [comment]
List of running docker at the server.

![image](https://user-images.githubusercontent.com/23035531/183419255-8edf28ee-8780-4ff4-8042-f8593756f8ff.png)

**Comment by Ahmedsaber9** [comment]
`docker run pospielov/compreface-core:1.0.0-mobilenet-2d106
[uWSGI] getting INI configuration from uwsgi.ini
*** Starting uWSGI 2.0.19 (64bit) on [Mon Aug  8 13:24:33 2022] ***
compiled with version: 10.2.1 20210110 on 10 April 2022 08:37:08
os: Linux-4.15.0-189-generic #200-Ubuntu SMP Wed Jun 22 19:53:37 UTC 2022
nodename: 9d3fd3642a5c
machine: x86_64
clock source: unix
detected number of CPU cores: 8
current working directory: /app/ml
detected binary path: /usr/local/bin/uwsgi
!!! no internal routing support, rebuild with pcre support !!!
setgid() to 33
setuid() to 33
your memory page size is 4096 bytes
detected max file descriptor number: 1048576
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uwsgi socket 0 bound to TCP address 0.0.0.0:3000 fd 3
Python version: 3.7.12 (default, Sep 28 2021, 19:53:04)  [GCC 10.2.1 20210110]
Python main interpreter initialized at 0x55bddc9102d0
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 218712 bytes (213 KB) for 2 cores
*** Operational MODE: preforking ***
{"severity": "INFO", "message": "{'BUILD_VERSION': 'dev', 'CALCULATION_PLUGIN': 'insightface.Calculator@arcface_mobilefacenet', 'EXTRA_PLUGINS': ['insightface.Landmarks2d106Detector', 'insightface.GenderDetector', 'insightface.AgeDetector', 'insightface.facemask.MaskDetector'], 'FACE_DETECTION_PLUGIN': 'insightface.FaceDetector@retinaface_mnet025_v1', 'GPU_IDX': -1, 'IMG_LENGTH_LIMIT': 640, 'INTEL_OPTIMIZATION': False, 'IS_DEV_ENV': False, 'LOGGING_LEVEL_NAME': 'DEBUG', 'ML_PORT': 3000}", "request": null, "logger": "src.app", "module": "app", "traceback": null, "build_version": "dev"}
{"severity": "DEBUG", "message": "Creating new app for WSGI", "request": null, "logger": "src.app", "module": "app", "traceback": null, "build_version": "dev"}
WSGI app 0 (mountpoint='') ready in 0 seconds on interpreter 0x55bddc9102d0 pid: 1 (default app)
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 1)
spawned uWSGI worker 1 (pid: 21, cores: 1)
spawned uWSGI worker 2 (pid: 22, cores: 1)`
![image](https://user-images.githubusercontent.com/23035531/183429240-99f80eff-d337-495b-b47f-4a87b6f89e34.png)

trying to curl with localhost : 

`word@word:~$ curl localhost:3000
curl: (7) Failed to connect to localhost port 3000: Connection refused
word@word:~$ curl 127.0.0.1:3000
curl: (7) Failed to connect to 127.0.0.1 port 3000: Connection refused`

![image](https://user-images.githubusercontent.com/23035531/183429135-2d5d7cd9-6b0e-492a-a092-94f56b99a08b.png)

**Comment by pospielov** [comment]
you need to stop CompreFace: `docker-compose down`
then replace image in docker-compose file, instead of:
```  compreface-core:
    restart: always
    image: ${registry}compreface-core:${CORE_VERSION}
    container_name: "compreface-core"
    environment:
      - ML_PORT=3000
```

it will be: 
```  compreface-core:
    restart: always
    image: pospielov/compreface-core:1.0.0-mobilenet-2d106
    container_name: "compreface-core"
    environment:
      - ML_PORT=3000
 ```
 
 then run CompreFace again

**Comment by Ahmedsaber9** [comment]
Worked like charm! 
I will push my contraption soon!
many thanks

**Comment by Ahmedsaber9** [comment]
Please note that after this update the admin panel is not working when you try to test with the GUI. 
I don't really need it, I'm just sharing it with you.
https://imgur.com/yoavVss

Thanks.

**Comment by Ahmedsaber9** [comment]
I'm curious about something, how far the used model is pre-trined? as I know you are using Google-Face net?
also after using the 1062d landmarks how can estimate the accuracy as % ?  did you benchmark that or something like this? 
also what if we increased the landmarks by more than 106? I saw some codes that get around 400 - 1000 landmarks.

Thanks.

**Comment by pospielov** [comment]
Yes, I know that GUI is broken, this is why it's not in the official distribution. 

1. We use pre-trained models from [FaceNet](https://github.com/davidsandberg/facenet) repository and from [InsightFace](https://github.com/deepinsight/insightface) repository. In the case of the 1062d image, it's the InsightFace model. You can find the accuracies here: https://github.com/exadel-inc/CompreFace/tree/master/custom-builds
2. That's a good question, I didn't come across the accuracies of landmark detectors in those libraries.
3. Everything is possible. But we don't have a big demand for even the 1062d feature. Also, it's easier to reuse the existing libraries. So as soon as InsightFace doesn't support it, it's harder to add it to CompreFace.
4. It only speeds up the process. 
5. Yes, I mean CompreFace is open sourced so all you need is to train the model, put it into the code and build the CompreFace. You can use official instructions on how to build the model from FaceNet and InsightFace

**Comment by dtheb** [comment]
Any update on this? since this is an old issue any chance there is there a new plugin or something to do liveness now?
